packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -11,6 +10,15 @@ import 'package:flutter/foundation.dart';
 import 'text_editing.dart';
 import 'text_input.dart';
 
+/// Signature for a callback function that will be applied on each banned
```

Comment:
Hmm I might be able to turn the callback to an abstract class and have it applied on both banned and allowed patterns, and tell the user to maintain their own state (to get rid of the public accumulator/mutable text range classes). 

I guess an enum works for existing use cases but wouldn't it be more extensible if we just expose the callback?

Reply:
I think we should have a clear idea why filterTextInputFormatter exist, if we make it too generic, user probably better off just implement the Textinputformatter abstract class.

Reply:
Oh touche. Yeah the base implementation isn't complicated, instead of learning "our way" of doing things people can just implement their own formatter.

Reply:
I am not sure whether this is cleaner. We can have strong opinion on how composing/selection works in filterTextInputFormatter, if user want more flexibility, they can implement the Textinputformatter base class. If there are really popular alternative way of doing the composing or selection, we can just create another Textinputformatter subclass to just handle that

Reply:
Ah I remember now: both strategies provided move the index to the "end" of the replacement, when the index is strictly inside a banned word.  People may want to do that differently.

Reply:
I would prefer a less opinionated implementation if it's not breaking, roughly equally performant, and does not have a steep learning curve. I think the public accumulator and `MutableTextRange` class are definitely confusing, let me see if I can reduce the API surface.

Reply:
Oh I can put the "strategy" stuff into a superclass of `FliteringTextInputFormatter`. But I don't have a good name for it, it's difficult to more get generic than `FliteringTextInputFormatter`.